provider/core/src/dynutil.rs

```diff
@@ -189,6 +204,52 @@ macro_rules! impl_dyn_provider {
             $crate::serde::SerializeMarker
         );
     };
+    ($provider:ty, [ $($struct_m:ty),+, ], SERDE_SE, impl DataConverter) => {
+        $crate::impl_dyn_provider!(
+            $provider,
+            [ $($struct_m),+, ],
+            SERDE_SE
+        );
+
+        // If this fails to compile, enable the "datagen" feature on this crate.
+        $crate::impl_dyn_provider!(
+            $provider,
+            { $(<$struct_m as $crate::ResourceMarker>::KEY => $struct_m),+, },
+            impl DataConverter
+        );
+    };
+    ($provider:ty, { $($pat:pat $(if $guard:expr)? => $struct_m:ty),+, }, impl DataConverter) => {
+        // If this fails to compile, enable the "datagen" feature on this crate.
+        impl $crate::datagen::DataConverter<$crate::any::AnyMarker, $crate::serde::SerializeMarker> for $provider {
+            fn convert(&self, key: $crate::ResourceKey, from: DataPayload<$crate::any::AnyMarker>) -> Result<$crate::DataPayload<$crate::serde::SerializeMarker>, $crate::datagen::ReturnedPayloadError<$crate::any::AnyMarker>> {
+                use $crate::datagen::ReturnedPayloadError;
+                match key {
+                    $(
+                        $pat $(if $guard)? => {
+                            let result: $crate::DataPayload<$struct_m> = from.downcast().expect("Downcasting failed!");
```

Comment:
Nit: Fix this to return a `ReturnedPayloadError` populated with the payload that failed to downcast

Reply:
This needs the downcasting methods to also do the same everywhere.

There are a bunch of places ReturnedPayloadError needs to be introduced, perhaps we can do it all at once in a separate PR?

Given that we're already checking the key type it's somewhat of a code bug if this branch is hit anyway so the expect is also somewhat correct.

Reply:
OK to do it in another PR.